src/ddsrt/include/dds/ddsrt/hopscotch.h

```diff
@@ -81,29 +226,194 @@ struct ddsrt_chh_iter {
 };
 #endif
 
+/**
+ * @brief Concurrent version of @ref ddsrt_hh_new
+ * 
+ * @param[in] init_size the minimum initial size
+ * @param[in] hash the hash function, see @ref ddsrt_hh_hash_fn
+ * @param[in] equals the equals function, see @ref ddsrt_hh_equals_fn
+ * @param[in] gc_buckets user defined garbage collection function, see @ref ddsrt_hh_buckets_gc_fn
+ * @param[in] gc_buckets_arg extra argument for @ref ddsrt_hh_buckets_gc_fn
+ * @return pointer to the hopscotch hash table
+ * @see @ref ddsrt_chh_free
+ */
 struct ddsrt_chh *ddsrt_chh_new (uint32_t init_size, ddsrt_hh_hash_fn hash, ddsrt_hh_equals_fn equals, ddsrt_hh_buckets_gc_fn gc_buckets, void *gc_buckets_arg);
+
+/**
+ * @brief Concurrent version of @ref ddsrt_hh_free
+ * 
+ * @param[in,out] hh the hash table to destroy
+ * @see @ref ddsrt_chh_new
+ */
 void ddsrt_chh_free (struct ddsrt_chh * __restrict hh);
+
+/**
+ * @brief Concurrent version of @ref ddsrt_hh_lookup
+ * 
+ * @param[in] rt the hash table
+ * @param[in] keyobject is the object with which to do the lookup
+ * @return pointer to the matching element, NULL if failed
+ */
 void *ddsrt_chh_lookup (struct ddsrt_chh * __restrict rt, const void * __restrict keyobject);
+
+/**
+ * @brief Concurrent version of @ref ddsrt_hh_add
+ * 
+ * @param[in,out] rt the hash table
+ * @param[in] data user data to add
+ * @return int meant to be interpreted as bool (1 is success, 0 is failure)
+ * @see @ref ddsrt_chh_remove
+ */
 int ddsrt_chh_add (struct ddsrt_chh * __restrict rt, void * __restrict data);
+
+/**
+ * @brief Concurrent version of @ref ddsrt_hh_remove
+ * 
+ * @param[in,out] rt the hash table
+ * @param[in] keyobject user data to remove
+ * @return int meant to be interpreted as bool (1 is success, 0 is failure)
+ * @see @ref ddsrt_chh_add
+ */
 int ddsrt_chh_remove (struct ddsrt_chh * __restrict rt, const void * __restrict keyobject);
+
+/**
+ * @brief Concurrent version of @ref ddsrt_hh_enum
+ * 
+ * Called unsafe because it doesn't use a mutex. The user must ensure no other thread is accessing the hash table in the mean time.
```

Comment:
It is not that another thread may not access the hash table. Concurrent lookups are self-evidently safe (because they don't alter the table), but adds and removes may also be performed while enumerating.

What is "unsafe" about it is that:
* if another thread is removing an element, then there is no guarantee that the element for which `f` is invoked is still in the table at the time of invocation (similar to a concurrent `remove` and `lookup`)
* if another thread is adding an element, there is no guarantee whether that new element will, or will not, be visited
* if another thread is adding an element, there is no guarantee that existing elements will be visited
* if another thread is adding an element, there is no guarantee that an element won't be visited multiple times

The first two are quite ordinary uncertainties. The latter two are nasty enough to make it deserve to be called "unsafe". I think it is better to be precise about this in the documentation.

Reply:
> It is not that another thread may not access the hash table. Concurrent lookups are self-evidently safe (because they don't alter the table), 

Wouldn't it still be problematic if say:
- Thr A calls ddsrt_chh_enum_unsafe with a free funtion.
- Whilst A's call is ongoing, Thr B does a lookup of element x.
- Thr A has freed element x.
- Thr B tries to read and modify data value of element x.

Reply:
Well, yes, but if the caller frees memory that it still uses, it's the callers fault 😀 It is true that this really does place a burden on the callers, see, e.g., https://github.com/eclipse-cyclonedds/cyclonedds/blob/1aeea0faea21ea14d587d46c02736c74d520440f/src/core/ddsi/src/ddsi_tkmap.c#L157 and https://github.com/eclipse-cyclonedds/cyclonedds/blob/1aeea0faea21ea14d587d46c02736c74d520440f/src/core/ddsi/src/ddsi_tkmap.c#L213